src/buf/fixed/plumbing/pool.rs

```diff
@@ -0,0 +1,186 @@
+        let mut free_buf_head_by_cap = HashMap::new();
+        for (index, mut buf) in bufs.enumerate() {
+            let cap = buf.capacity();
+
+            // Link the buffer as the head of the free list for its capacity.
+            // This constructs the free buffer list to be initially retrieved
+            // back to front, which should be of no difference to the user.
+            let next = free_buf_head_by_cap.insert(cap, index as u16);
+
+            iovecs.push(iovec {
+                iov_base: buf.as_mut_ptr() as *mut _,
+                iov_len: cap,
+            });
+            states.push(BufState::Free {
+                init_len: buf.len(),
+                next,
+            });
+            mem::forget(buf);
+        }
+        debug_assert_eq!(iovecs.len(), states.len());
+
+        // Safety: Vec::as_mut_ptr never returns null
+        let raw_bufs = unsafe { ptr::NonNull::new_unchecked(iovecs.as_mut_ptr()) };
+        let orig_cap = iovecs.capacity();
+        mem::forget(iovecs);
+        Pool {
+            raw_bufs,
+            orig_cap,
+            states,
+            free_buf_head_by_cap,
+            waiting_on_next: vec![],
+        }
+    }
+
+    // If the free buffer list for this capacity is not empty, checks out the first buffer
+    // from the list and returns its data. Otherwise, returns None.
+    pub(crate) fn try_next(&mut self, cap: usize) -> Option<CheckedOutBuf> {
+        let free_head = self.free_buf_head_by_cap.get_mut(&cap)?;
+        let index = *free_head as usize;
+        let state = &mut self.states[index];
+
+        let (init_len, next) = match *state {
+            BufState::Free { init_len, next } => {
+                *state = BufState::CheckedOut;
+                (init_len, next)
+            }
+            BufState::CheckedOut => panic!("buffer is checked out"),
+        };
+
+        // Update the head of the free list for this capacity.
+        match next {
+            Some(i) => {
+                *free_head = i;
+            }
+            None => {
+                self.free_buf_head_by_cap.remove(&cap);
+            }
+        }
+
+        // Safety: the allocated array under the pointer is valid
+        // for the lifetime of self, a free buffer index is inside the array,
+        // as also asserted by the indexing operation on the states array
+        // that has the same length.
+        let iovec = unsafe { self.raw_bufs.as_ptr().add(index).read() };
+        debug_assert_eq!(iovec.iov_len, cap);
+        Some(CheckedOutBuf {
+            iovec,
+            init_len,
+            index: index as u16,
+        })
+    }
+
+    pub(crate) fn poll_next(&mut self, cap: usize, cx: &mut Context<'_>) -> Poll<CheckedOutBuf> {
+        if let Some(buf) = self.try_next(cap) {
+            return Poll::Ready(buf);
+        }
+        let waker = cx.waker();
+        if !self.waiting_on_next.iter().any(|w| w.will_wake(waker)) {
```

Comment:
Sounds like will_wake won't catch all duplicates. Is it possible this list will grow too large? And what's the danger of keeping an old waker that the scheduler had replaced in a subsequent poll?

Reply:
> Sounds like will_wake won't catch all duplicates. Is it possible this list will grow too large?

It is theoretically, but I don't know a way to prevent this.

> And what's the danger of keeping an old waker that the scheduler had replaced in a subsequent poll?

I trust the `Waker` API to do the right thing in this case.

Reply:
Tokio still has an open issue about this. https://github.com/rust-lang/rust/issues/66281

How about keeping an optional clone in the future itself, to allow searching the `Vec` with the older one first, with simple address comparison? I'm afraid the `Vec`'s growth could be unbounded if the future were part of a select in a loop as long as Tokio's waker API doesn't guarantee will_wake doesn't have false negatives.

Reply:
> How about keeping an optional clone in the future itself, to allow searching the `Vec` with the older one first, with simple address comparison?

How could that work with cloned wakers that do not expose any underlying pointers in stable API?

Reply:
If you store a waker in the Future, don't you have a stable address you can compare against?

Reply:
No. The reference returned by `cx.waker()` is not guaranteed to live longer than the poll call that the `Context` is passed into. Cloning the waker is possible, but what address comparison against other `Waker` instances is going to achieve then? The only API we have to avoid storing redundant wakers is `Waker::will_wake`. That's not guaranteed to always work, but I think in practice no executor will repeatedly produce wakers that can't be correlated to one another with `will_wake` when the task is repeatedly polled on a pending future. Correct me please if there are examples to the contrary.

Reply:
I'm sorry I don't see the problem yet with storing a cloned waker in the future when the cloned waker is also pushed onto the `Vec` so when poll is called again, before pushing a new cloned waker onto the `Vec`, the cloned waker held by the future is used to get the old waker's address and that address is searched in the `Vec` to see if a `Vec` location can be swapped out with a new clone. I have not seen any structures in the std library where a list of wakers is held. I'm prepared to write a demo to see if this idea would work, I don't see why it wouldn't - the waker is a function and it has an address that can be compared, but I haven't had time yet.

I know `will_wake` is used for the first best shot at deciding when a waker doesn't need to be saved because an equally valid waker is already held. But in all the cases I've seen, when `will_wake` returns false, it means a waker is replaced with a clone, not appended as a clone. I'm just trying to find the location in the `Vec` to perform the replacement by comparing pointers. Another approach, not relying on the waker address, would be to store a generation counter along with the waker in the `Vec` cell and keep the generation counter in the future, for searching the `Vec` on a subsequent poll.

Reply:
I have rewritten the wakeup implementation using `tokio::sync::Notify`. It's non-allocating, besides the synchronization objects created on demand for each capacity bucket.